src/arithmetic.cpp

```diff
@@ -1806,6 +1763,52 @@ void __mul_ex(fp* z, const fp* x, const fp* y)
     asm("pop %r15;");
     asm("pop %rbp;");
 }
+
+typedef void (*blsmul_func_t)(fp*, const fp*, const fp*);
+
+static bool cpu_has_bmi2_and_adx() {
+    // borrowed from: https://github.com/Mysticial/FeatureDetector/blob/master/src/x86/cpu_x86.cpp
+    int32_t info[4];
+    __cpuid_count(0, 0, info[0], info[1], info[2], info[3]);
+    int nIds = info[0];
+    if(nIds >= 0x00000007) {
+        __cpuid_count(0x00000007, 0, info[0], info[1], info[2], info[3]);
+        if((info[1] & (1 <<  8)) && (info[1] & (1 << 19))) // BMI2 && ADX
+            return true;
+    }
+    return false;
+}
+
+#ifdef __ELF__
+extern "C" char** _dl_argv;
+
+extern "C" blsmul_func_t __attribute__((no_sanitize_address)) resolve_blsmul() {
+    int argc = *(int*)(_dl_argv - 1);
+    char** my_environ = (char**)(_dl_argv + argc + 1);
+    while(*my_environ != nullptr) {
+       const char disable_str[] = "BLS_DISABLE_BMI2";
+        if(strncmp(*my_environ++, disable_str, strlen(disable_str)) == 0)
+           return __mul;
+    }
+
+    if(cpu_has_bmi2_and_adx())
+       return __mul_ex;
+    return __mul;
+}
+
+void _mul(fp*, const fp*, const fp*) __attribute__((ifunc("resolve_blsmul")));
+#else
```

Comment:
You can add a link to IFUNC description https://gcc.gnu.org/onlinedocs/gcc-5.3.0/gcc/Function-Attributes.html#index-g_t_0040code_007bifunc_007d-function-attribute-3095 for future code maintainers.